Simplenote/SimplenoteAppDelegate+Swift.swift

```diff
@@ -8,8 +8,52 @@ import CoreSpotlight
 extension SimplenoteAppDelegate {
 
     @objc
-    func makeCoreDataManager() throws -> CoreDataManager {
-        try CoreDataManager()
+    func setupStorage() {
+        let migrationResult = SharedStorageMigrator().performMigrationIfNeeded()
+
+        do {
+            try setupCoreData(migrationResult: migrationResult)
+        } catch {
+            fatalError(error.localizedDescription)
+        }
+    }
+
+    private func setupCoreData(migrationResult: MigrationResult) throws {
+        let settings = StorageSettings()
+
+        switch migrationResult {
+        case .notNeeded, .success:
+            try validateStorageDirectory(at: settings.sharedUserLibraryDirectory)
+            coreDataManager = try CoreDataManager(at: settings.sharedStorageURL)
+        case .failed:
+            try validateStorageDirectory(at: settings.legacyUserLibraryDirectory)
+            coreDataManager = try CoreDataManager(at: settings.legacyStorageURL)
+        }
+    }
+
+    private func validateStorageDirectory(at url: URL) throws {
+        // Validate the directory for the store DB
+        do {
+            try validateResourceValueForDirectory(at: url)
+        } catch {
+            try handleDirectoryError((error as NSError), directoryURL: url)
+        }
+    }
+
+    private func validateResourceValueForDirectory(at url: URL) throws {
+        let properties = try url.resourceValues(forKeys: [URLResourceKey.isDirectoryKey])
+
+        if properties.isDirectory != true {
+            throw CoreDataManagerError.foundNotDirectoryAtFilesDirectoryURL
+        }
+    }
+
+    private func handleDirectoryError(_ error: NSError, directoryURL: URL) throws {
+        if error.code == NSFileReadNoSuchFileError {
+            try FileManager.default.createDirectory(at: directoryURL, withIntermediateDirectories: true, attributes: nil)
+        } else {
+            throw error
+        }
     }
```

Comment:
These 3 last API(s) can probably be simplified as follows **(untested)**:

```
extension FileManager {

    func ensureDirectoryExists(at url: URL) throws {
        if directoryExistsAtURL(url) {
            return
        }

        try createDirectory(at: url, withIntermediateDirectories: true)
    }

/// This can prob. be modernized, borrowed from elsewhere!
    func directoryExistsAtURL(_ url: URL) -> Bool {
        var isDir: ObjCBool = false
        let exists = fileExists(atPath: url.path, isDirectory: &isDir)
        return exists && isDir.boolValue
    }
}
```